File: DSA/src/Questions/Array/WaysToPartition.java
package Questions.Array;

import java.util.HashMap;
import java.util.Map;
/*
https://leetcode.com/problems/maximum-number-of-ways-to-partition-an-array/description/
 You are given a 0-indexed integer array nums of length n. The number of ways to partition nums is the number of pivot indices that satisfy both conditions:

1 <= pivot < n
nums[0] + nums[1] + ... + nums[pivot - 1] == nums[pivot] + nums[pivot + 1] + ... + nums[n - 1]
You are also given an integer k. You can choose to change the value of one element of nums to k, or to leave the array unchanged.

Return the maximum possible number of ways to partition nums to satisfy both conditions after changing at most one element.

 

Example 1:

Input: nums = [2,-1,2], k = 3
Output: 1
Explanation: One optimal approach is to change nums[0] to k. The array becomes [3,-1,2].
There is one way to partition the array:
- For pivot = 2, we have the partition [3,-1 | 2]: 3 + -1 == 2.
Example 2:

Input: nums = [0,0,0], k = 1
Output: 2
Explanation: The optimal approach is to leave the array unchanged.
There are two ways to partition the array:
- For pivot = 1, we have the partition [0 | 0,0]: 0 == 0 + 0.
- For pivot = 2, we have the partition [0,0 | 0]: 0 + 0 == 0.
Example 3:

Input: nums = [22,4,-25,-20,-15,15,-16,7,19,-10,0,-13,-14], k = -33
Output: 4
Explanation: One optimal approach is to change nums[2] to k. The array becomes [22,4,-33,-20,-15,15,-16,7,19,-10,0,-13,-14].
There are four ways to partition the array.
 

Constraints:

n == nums.length
2 <= n <= 105
-105 <= k, nums[i] <= 105
 */
public class WaysToPartition {
    public int waysToPartition(int[] nums, int k) {

        long s = 0;
        for (int n : nums)
            s += n;
        Map<Long, Integer> map = new HashMap<Long, Integer>();
        long r = 0;
        for (int i = nums.length - 1; i > 0; i--) {
            r += nums[i]; // sufix sum from i
            long t = s - r * 2 + k; // this suffix sum (r) would qualify as pivot if in the corresponding prefix an
                                    // element of value t was changed to k
            map.put(t, map.getOrDefault(t, 0) + 1);
        }

        long l = 0;
        int resp = map.getOrDefault((long) nums[0], 0);
        int none = 0; // count of pivots for unchanged array
        for (int i = 1; i < nums.length; i++) {
            l += nums[i - 1]; // prefix sum up to i-1
            if (l * 2 == s)
                none++; // i is a pivot for unchanged array
            long t = s - l * 2;
            map.put(t + k, map.getOrDefault(t + k, 0) + 1); // this prefix sum (l) would qualify as pivot if in the
                                                            // corresponding suffix an element of value t+k was changed
                                                            // to k
            map.put(-t + k, map.get(-t + k) - 1); // remove the suffix contribution
            resp = Math.max(resp, map.getOrDefault((long) nums[i], 0)); // check the
        }
        return Math.max(resp, none);
    }

    public static void main(String[] args) {
        WaysToPartition obj = new WaysToPartition();
        int[] nums = { 2, -1, 2 };
        int k = 3;
        System.out.println(obj.waysToPartition(nums, k));
    }
}

File: DSA/Python/Questions/Array/WaysToPartition.py
"""
# Question: Maximum Number of Ways to Partition an Array
# Link: https://leetcode.com/problems/maximum-number-of-ways-to-partition-an-array/

# Find max ways to partition array after changing at most one element to k

# Time Complexity: O(n) where n is length of array
# Space Complexity: O(n) for storing prefix sums

# Algorithm:
# 1. Calculate total sum and build prefix/suffix maps
# 2. Track partition points with and without changes
# 3. Handle both unchanged array and single element changes
# 4. Update counts based on prefix and suffix sums

# Key Components:
# - ways_to_partition(): Main function implementing partition counting
# - HashMap for sum frequency tracking
# - Dynamic prefix/suffix handling
"""

from typing import List
from collections import defaultdict


class WaysToPartition:
    def ways_to_partition(self, nums: List[int], k: int) -> int:
        total_sum = sum(nums)
        sum_map = defaultdict(int)
        right_sum = 0

        # Build suffix map
        for i in range(len(nums) - 1, 0, -1):
            right_sum += nums[i]
            target = total_sum - right_sum * 2 + k
            sum_map[target] += 1

        left_sum = 0
        response = sum_map.get(nums[0], 0)
        unchanged_count = 0

        # Process each position
        for i in range(1, len(nums)):
            left_sum += nums[i - 1]
            if left_sum * 2 == total_sum:
                unchanged_count += 1

            target = total_sum - left_sum * 2
            sum_map[target + k] = sum_map.get(target + k, 0) + 1
            sum_map[-target + k] = sum_map.get(-target + k, 0) - 1
            response = max(response, sum_map.get(nums[i], 0))

        return max(response, unchanged_count)


def main():
    solution = WaysToPartition()
    nums = [2, -1, 2]
    k = 3
    result = solution.ways_to_partition(nums, k)
    print(result)  # Expected output: 1


if __name__ == "__main__":
    main()
